### ISSUE TITLE
Entity Collection Loader not returning expected result

### ISSUE THREAD

**Issue by DanJBower** [corstian/Boerman.GraphQL.Contrib#12]
Do you have any examples of using `EntityCollectionLoader`?

I'm having an issue with using it to load data out of a junction table. To try and load data out the table, I am using it like below. I think I have written it so it loads the pupils in the class that has the ID `context.Source.Id`.

	IList<Pupils> list = (await accessor.EntityCollectionLoader(
		() => schoolContext.ClassesPupils,
		classesPupils => classesPupils.ClassId,
		select => select.Pupil,
		pupil => pupil.Id,
		context.Source.Id)
			.ConfigureAwait(false)).ToList();

I tried debugging your source and upon inspection, it found the 11 pupils I expected it to find.

![Source](https://user-images.githubusercontent.com/40743213/72734363-6dffd180-3b91-11ea-865a-892eedaf39e9.png)

However, the return of the function results in an empty enumerable.

![Empty List](https://user-images.githubusercontent.com/40743213/72734558-e23a7500-3b91-11ea-966c-45689743cabe.png)

Do you have any idea why this might be?

The basic data structure I am using look like the below:

	public sealed class Classes : IClass
    {
        public Classes()
        {
            ClassesPupils = new HashSet<ClassesPupils>();
        }
		
        public Guid Id { get; set; }

        public Guid RoomId { get; set; }

        public Classrooms Room { get; set; }

        public ICollection<ClassesPupils> ClassesPupils { get; set; }

        public IEnumerable<IPupil> Pupils => ClassesPupils.Select(_ => _.Pupil);

        public IClassroom Classroom => Room;
    }
	
	public sealed class ClassesPupils
    {
        public Guid ClassId { get; set; }

        public Guid PupilId { get; set; }

        public Classes Class { get; set; }

        public Pupils Pupil { get; set; }
    }

	public sealed class ClassType : ObjectGraphType<Classes>
	{
		public ClassType(IDataLoaderContextAccessor accessor, ISchoolContext schoolContext)
		{
			Field(@class => @class.Id, type: typeof(IdGraphType)).Description("The class ID");
			
			Field(@class => @class.Classroom, type: typeof(ClassroomType)).Description("The classroom the class is in.")
				.ResolveAsync(
					async context =>
					{
						return await accessor.EntityLoader(
							() => schoolContext.Classrooms,
							classroom => classroom.Id,
							context.Source.RoomId)
								.ConfigureAwait(false);
					}); // This works great
					
			Field(@class => @class.Pupils, type: typeof(ListGraphType<PupilType>)).Description("The pupils in a class")
				.ResolveAsync(
					async context =>
					{
						IList<Pupils> list = (await accessor.EntityCollectionLoader(
		                                       () => schoolContext.ClassesPupils,
		                                       classesPupils => classesPupils.ClassId,
		                                       select => select.Pupil,
		                                       pupil => pupil.Id,
		                                       context.Source.Id)
			                                   .ConfigureAwait(false)).ToList();
						return list; // List is empty here
					});
		}
	}

**Comment by corstian** [comment]
The main difference between the entity loader and the collection loader is the `Task.FromResult` return in the latter. However strange this is but I suspect that the `GetOrAddCollectionBatchLoader` does not know how to deal with non async methods, and as such does not properly return the expected result.

At the moment I don't really have any code to test this hypothesis on, but I can give it a shot somewhere in the upcoming days.

Long story short, adding the async keyword and removing the `Task.FromResult` return might solve the issue.

**Comment by DanJBower** [comment]
Unfortunately, I couldn't get it to work by adding `async` and removing `Task.FromResult`.

I appreciate the suggestion and you looking in to it 😃

**Comment by DanJBower** [comment]
Hi @corstian 

I have created a small repository that exactly recreates my issue.

[BasicBatchCollectionTest](https://github.com/DanJBower/BasicBatchCollectionTest)

The repository has a sql_dump.xml that allows you to restore an example sql database schema with some sample data in.

If you run the `BasicCollectionLoaderTest` project and enter the query

	{
	  classes {
	    id,
	    subject,
	    pupils {
	      id,
	      name
	    }
	  }
	}

It will return 

    {
      "data": {
        "classes": [
          {
            "id": "0273bef8-040c-4f1a-b3fd-b3db00cc16a0",
            "subject": "Technology",
            "pupils": []
          },
          {
            "id": "2655d463-fce4-4b37-8d5a-fdca97d79b9a",
            "subject": "Maths",
            "pupils": []
          },
          {
            "id": "383dc8ce-02fd-49c2-8be5-30fe6cd569e9",
            "subject": "Engineering",
            "pupils": []
          },
          {
            "id": "b057e19b-50ad-4690-b8eb-070486c21655",
            "subject": "Science",
            "pupils": []
          }
        ]
      }
    }

Whereas, I would expect pupils for each of the classes to be populated as `ClassType` defines how pupils should be resolved. If you do run it, you will probably have to update the details in `SchoolContext.Password` as it currently has a dummy value in.

I have included an extract of the method from your source but with the suggested changes you made in `SourceCopy\Dataloader.cs`

Thanks again for the help.

**Comment by corstian** [comment]
Whoa this was a difficult one to track down. Thank you for providing a reproduction. So here is the thing;

With every `GraphObjectType` which makes use of the `EntityCollectionLoader` (in your case the `ClassType`), the calling type has to provide an id for which related objects need to be retrieved.

However, the `EntityCollectionLoader` and more importantly the underlying `GetOrAddCollectionBatchLoader` method, do not know about what retrieved entities are related to which ids you have provided. In order to create that connection an expression needs to provided.

I think this error could be prevented in the future by either adding proper documentation to the method, or logging a warning when the returned entries in the `Lookup` type are not equal to the provided items (keys).

The signature of the `EntityCollectionLoader` can be explained as follows:

```csharp
public static async Task<IEnumerable<TSelect>> EntityCollectionLoader<T, TValue, TSelect>(
            this IDataLoaderContextAccessor dataLoader,
            Func<IQueryable<T>> queryableAccessor,
            Expression<Func<T, TValue>> dataPredicate,
            Expression<Func<T, TSelect>> selector,
            Expression<Func<TSelect, TValue>> outputPredicate,
            TValue value)
            where T : class
```

`this IDataLoaderContextAccessor dataLoader`: the graph data loader itself
`Func<IQueryable<T>> queryableAccessor`: The data source where to query from
`Expression<Func<T, TValue>> dataPredicate`: Selects the key from the data source which needs to match to any in the requested keys (of type `TValue`, in your case the `ClassId`)
`Expression<Func<T, TSelect>> selector`:  Defines which data will be returned
`Expression<Func<TSelect, TValue>>` Defines how to match the return data (selected above) with the key provided.
`TValue value`: The key which is provided

In order to solve your problem you might be better off to query your connection table instead:

```csharp
.ResolveAsync(
    async context =>
    {
        var pupils = await accessor.EntityCollectionLoader(
            () => schoolContext.ClassPupil.Include(q => q.Pupil),
            q => q.ClassId,
            q => q,
            q => q.ClassId,
            context.Source.Id);

        return pupils.Select(q => q.Pupil);
    })
```

While it seems a bit abstract, first you query the ClassPupil table, and tell EF Core to include the Pupil entity in the result. After that you tell the data loader to use the ClassId field to match the queried value. Because the same value is selected as is queried by the data source (the `IQueryable`), the third expression also queries the same field as the first one.

I am pushing a PR to your branch with a working version. Would you mind if I link your branch as an example for usage of the data loaders?

Let me know whether this solves your questions/problems! Happy hacking!

----

Ps. as you started using data loaders I would recommend using the following structure to build your fields:

```csharp
Field<ListGraphType<PupilType>>()
    .Name("pupils")
    .Description("Pupils enrolled in this class")
    .ResolveAsync(async context =>
    {
        throw new NotImplementedException();
    });
```

Doing this makes it more clear what the returned type is, and there is less confusion about whether the `_ => _.Pupils` provide the return value or whether the `ResolveAsync` method does.

**Comment by DanJBower** [comment]
Thanks a lot for the help. I appreciate it.

I've merged your PR and your tip is helpful :)

Also, you are welcome to link the branch for any examples.

Thanks again.
